### PR TITLE
fix(events): Revisit reflecting application paused/resumed events on resources

### PR DESCRIPTION
**Pull request by luispollo**
This is an alternative take on #842 and #853. Instead of injecting synthetic resource events in the history, we inject the original application-level paused/resumed events as appropriate.

Sample API response:
```yaml
- type: "ApplicationActuationPaused"
  application: "fnord"
  reason: "Application fnord paused"
  timestamp: "2020-03-11T06:27:03.453Z"
  scope: "application"
  uid: "fnord"
- type: "ApplicationActuationResumed"
  application: "fnord"
  reason: "Application fnord resumed"
  timestamp: "2020-03-11T06:17:03.453Z"
  scope: "application"
  uid: "fnord"
- type: "ApplicationActuationPaused"
  application: "fnord"
  reason: "Application fnord paused"
  timestamp: "2020-03-11T06:07:03.453Z"
  scope: "application"
  uid: "fnord"
- type: "ResourceValid"
  kind: "test/whatever@v1"
  id: "test:whatever:33323037"
  application: "fnord"
  timestamp: "2020-03-11T05:57:03.453Z"
  scope: "resource"
  uid: "test:whatever:33323037"
- type: "ResourceCreated"
  kind: "test/whatever@v1"
  id: "test:whatever:33323037"
  application: "fnord"
  timestamp: "2020-03-11T05:47:03.453Z"
  ignoreRepeatedInHistory: false
  scope: "resource"
  uid: "test:whatever:33323037"
```

### CODE REVIEW

**Comment by erikmunson** [comment]
nice.

tiny tactical request for my own convenience on the UI — can you also include the `scope` thing you were modeling on these events in previous PRs? e.g. the resource events get `scope: 'resource'`, app events get `scope: 'application'`?

makes it much easier for me to express the types on the client without handwaving some optional fields around.

**Comment by luispollo** [comment]
> nice.
> 
> tiny tactical request for my own convenience on the UI — can you also include the `scope` thing you were modeling on these events in previous PRs? e.g. the resource events get `scope: 'resource'`, app events get `scope: 'application'`?
> 
> makes it much easier for me to express the types on the client without handwaving some optional fields around.

Ah, yes! I thought of adding it and forgot. Now it makes sense to have it in the JSON.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt [hunk 1]
```diff
@@ -66,59 +69,55 @@ class ActuationPauser(
...
-      // We do this dynamically here so that it applies to all resources in the app, even those added _after_ the
-      // application was paused.
-      val appPausedEvents = resourceRepository
-        .applicationEventHistory(resource.application, events.last().timestamp)
-        .filterIsInstance<ApplicationActuationPaused>()
+  /**
+   * Adds [ApplicationActuationPaused] and [ApplicationActuationResumed] events to the resource event history so that
+   * it reflects actuation pauses/resumes at the application level.
+   */
+  fun addApplicationActuationEvents(originalEvents: List<ResourceEvent>, resource: Resource<*>) =
```
Comment (luispollo) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt]: Little picture if it helps...
image

**Comment by robfletcher** [comment]
Not started reviewing yet, but I did notice the other day that the `ignoreRepeatedInHistory` is appearing in JSON/YAML output which seems… unnecessary.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt [hunk 2]
```diff
@@ -15,7 +27,7 @@ abstract class PersistentEvent {
   }
 
   enum class Scope {
-    RESOURCE,
-    APPLICATION
+    @JsonProperty("resource") RESOURCE,
+    @JsonProperty("application") APPLICATION
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt]: What does this do? Those don't look like properties to me, how is that annotation relevant?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt [hunk 3]
```diff
@@ -1,8 +1,20 @@
...
+import com.fasterxml.jackson.annotation.JsonSubTypes
+import com.fasterxml.jackson.annotation.JsonTypeInfo
 import java.time.Clock
 import java.time.Instant
 
+@JsonTypeInfo(
+  use = JsonTypeInfo.Id.NAME,
+  property = "type",
+  include = JsonTypeInfo.As.PROPERTY
+)
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt]: isn't the discriminator actually the `scope` property? (in which case it would also be `include = As.EXISTING_PROPERTY`)

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt [hunk 4]
```diff
@@ -66,59 +69,55 @@ class ActuationPauser(
...
-        publisher.publishEvent(ResourceActuationResumed(it))
-      }
+    publisher.publishEvent(ApplicationActuationResumed(application, "Application $application resumed", clock))
   }
 
   fun pauseResource(id: String) {
     log.info("Pausing resource $id")
     pausedRepository.pauseResource(id)
-    publisher.publishEvent(ResourceActuationPaused(resourceRepository.get(id)))
+    publisher.publishEvent(ResourceActuationPaused(resourceRepository.get(id), null, clock.instant()))
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt]: What's the `null` there?

It would also be nice to be consistent about how the exception constructors use `Clock` vs `Instant`. The calls just above to `ApplicationActuationPaused` are passing in the clock, and here it's passing an instant. I don't have a strong preference (lean toward the latter) but it would be good if they always did the same thing.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt [hunk 5]
```diff
@@ -66,59 +69,55 @@ class ActuationPauser(
   fun pauseApplication(application: String) {
     log.info("Pausing application $application")
     pausedRepository.pauseApplication(application)
-    publisher.publishEvent(ApplicationActuationPaused(application))
+    publisher.publishEvent(ApplicationActuationPaused(application, "Application $application paused", clock))
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt]: Seems like that message could be generated inside the constructor rather than having to be passed in since it looks like it's always going to be the same

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt [hunk 6]
```diff
@@ -66,59 +69,55 @@ class ActuationPauser(
...
+          .filter { it is ApplicationActuationPaused || it is ApplicationActuationResumed }
 
-      appPausedEvents.forEach { appPaused ->
-        val lastBeforeAppPaused = events.firstOrNull { event ->
-          event.timestamp.isBefore(appPaused.timestamp)
-        }
+        relevantAppEvents.forEach { appEvent ->
+          val lastBeforeAppEvent = events.firstOrNull { event ->
+            event.timestamp.isBefore(appEvent.timestamp)
+          }
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt]: Couldn't we just apply `sortedBy { it.timestamp }` after we have the list rather than having to work out where to insert things?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt (hunk 1 above)
Comment (robfletcher) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt]: For my understanding… is this happening on-the-fly when we generate the list of events for the history API response?

Would it make sense to just make this a responsibility of the repository method that gets the events? It could do this with a single join query.

Comment (luispollo) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt]: It is. I guess we could change the query (no join required, it's the same `event` table), but I have a feeling that we might still go back and change this behavior (yet again), so maybe not worth it?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt (hunk 2 above)
Comment (luispollo) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt]: For enum values, this just uses the value you pass in as the enum value. In other words, just a way to make them lowercase here.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt (hunk 3 above)
Comment (luispollo) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt]: It is, but it turns out that is not what `property` is used for by Jackson. It's just used to name the JSON property that will include the sub-type `name` as defined below. I was confused by this too until I read the docs again.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt (hunk 5 above)
Comment (luispollo) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt]: Good catch, will fix.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt (hunk 4 above)
Comment (luispollo) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt]: The `null` is for `reason`. It was already `null` previously but the new constructor required it. As for clock vs. instant, I agree. The reason this case was "special" was that I was once creating synthetic `ResourceActuationPaused` events from an `ApplicationActuationPaused` by copying the timestamp. I can undo this now that we're just using the original application events.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt (hunk 6 above)
Comment (luispollo) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt]: Er... That would make total sense, wouldn't it? 🤦‍♂ Will change.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt (hunk 3 above)
Comment (luispollo) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt]: Having said that, I did not read up on, or try, `As.EXISTING_PROPERTY`. Let me give it a try.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt (hunk 2 above)
Comment (luispollo) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt]: From the docs:
> Starting with Jackson 2.6 this annotation may also be used to change serialization of Enum like so:
>
> ```java
> public enum MyEnum {
>     @JsonProperty("theFirstValue") THE_FIRST_VALUE,
>     @JsonProperty("another_value") ANOTHER_VALUE;
> }
> ```
> as an alternative to using `JsonValue` annotation.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt (hunk 3 above)
Comment (luispollo) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt]: I tried `AS.EXISTING_PROPERTY` with `scope` but that didn't work. For some reason it only works if I use `"type"` as the name of the property, which is the same name used in the `@JsonTypeInfo` annotation in the 2nd-level base classes. 🤷‍♂  

The reason I even added this here was so that Jackson could rebuild the class hierarchy (e.g. `PersistentEvent` -> `ResourceEvent` -> `ResourceCreated`), even though those base classes (which are, in fact, abstract) are never (de-)serialized themselves.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt (hunk 1 above)
Comment (robfletcher) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt]: yeah. I'm happy that it's on-the-fly. Let's go with it for a bit and we can consider optimizing later.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt (hunk 2 above)
Comment (robfletcher) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt]: That's great! We should change all the other places we're doing lowercasing the other way. I'll submit a PR

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt (hunk 3 above)
Comment (robfletcher) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt]: Aren't we going to end up with both `type` and `scope` in the JSON holding the same value?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/ApplicationEvent.kt [hunk 7]
```diff
@@ -36,9 +35,9 @@ data class ApplicationActuationPaused(
   @JsonIgnore
   override val ignoreRepeatedInHistory = true
 
-  constructor(application: String, reason: String? = null, clock: Clock = Companion.clock) : this(
+  constructor(application: String, clock: Clock = Companion.clock) : this(
     application,
-    reason,
+    "Application $application paused",
```
Comment (lorin) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/ApplicationEvent.kt]: What's the role of the reason field?

The comment on line 28 describes the reason field as `The reason why actuation was paused`, but the string being passed `"Application $application paused"` doesn't seem to provide additional information.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt (hunk 1 above)
Comment (lorin) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/pause/ActuationPauser.kt]: These diagrams are _really_ helpful, thanks for taking the time to generate them.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt (hunk 3 above)
Comment (luispollo) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt]: No, because `type` is "overwritten" by the concrete sub-classes (which are the only ones that can actually be created and hence serialized). The output looks like the sample in the PR description.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/ApplicationEvent.kt (hunk 7 above)
Comment (luispollo) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/ApplicationEvent.kt]: Yeah, fair point. I think I can just remove them.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt (hunk 3 above)
Comment (robfletcher) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt]: boy I was confused there. Looking at the sample it's clear they're different things
